src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs

```diff
@@ -1015,6 +1015,296 @@ public void GetAsync_ServerNeedsAuthAndNoCredential_StatusCodeUnauthorized()
             }
         }
 
+        [Theory]
+        [InlineData(":")]
+        [InlineData("  :  ")]
+        [InlineData("\x1234: \x5678")]
+        [InlineData("nocolon")]
+        [InlineData("no colon")]
+        [InlineData("Content-Length      ")]
+        public async Task GetAsync_InvalidHeaderNameValue_ThrowsHttpRequestException(string invalidHeader)
+        {
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                using (var client = CreateHttpClient())
+                {
+                    await Assert.ThrowsAsync<HttpRequestException>(() => client.GetStringAsync(uri));
+                }
+            }, server => server.AcceptConnectionSendCustomResponseAndCloseAsync($"HTTP/1.1 200 OK\r\nContent-Length: 11\r\n{invalidHeader}\r\n\r\nhello world"));
+        }
+
+        [Fact]
+        public async Task GetAsync_ManyDifferentRequestHeaders_SentCorrectly()
+        {
+            if (IsWinHttpHandler)
+            {
+                // Issue #TODO_BEFORE_MERGE
+                // Fails consistently with:
+                // System.InvalidCastException: "Unable to cast object of type 'System.Object[]' to type 'System.Net.Http.WinHttpRequestState'"
+                // This appears to be due to adding the Expect: 100-continue header, which causes winhttp
```

Comment:
WinHttpHandler (WinHTTP) doesn't support 'Expect: 100-continue" semantics.

Reply:
I know, but it shouldn't fail with an InvalidCastException after attempting to access the wrong state from a GCHandle :)